pulp_rpm/app/models.py

```diff
@@ -824,30 +1053,147 @@ class Environment(Content):
             A dictionary of names by language
         digest (Text):
             A checksum for the environment
+        packagegroups (ManyToMany):
+            PackageGroups related to this environment
+        optionalgroups (ManyToMany):
+            PackageGroups optionally related to this environment
     """
 
-    TYPE = 'environment'
+    TYPE = 'packageenvironment'
 
     # Required metadata
     id = models.CharField(max_length=255)
 
     name = models.CharField(max_length=255)
-    description = models.TextField()
-    display_order = models.IntegerField()
+    description = models.TextField(default='')
+    display_order = models.IntegerField(null=True)
 
     group_ids = JSONField(default=list)
     option_ids = JSONField(default=list)
 
-    desc_by_lang = models.TextField()
-    name_by_lang = models.TextField()
+    desc_by_lang = JSONField(default=dict)
+    name_by_lang = JSONField(default=dict)
 
     digest = models.CharField(unique=True, max_length=64)
 
+    packagegroups = models.ManyToManyField(PackageGroup, related_name='packagegroups_to_env')
+    optionalgroups = models.ManyToManyField(PackageGroup, related_name='optionalgroups_to_env')
+
     class Meta:
         default_related_name = "%(app_label)s_%(model_name)s"
 
+    @classmethod
+    def natural_key_fields(cls):
+        """
+        Digest is used as a natural key for PackageEnvironment.
+        """
+        return ('digest',)
+
+    @classmethod
+    def idlist_to_lst(cls, value):
+        """
+        Convert libcomps IdList to list.
+
+        Args:
+            value: a libcomps IdList
+
+        Returns:
+            A list
+
+        """
+        if len(value):
+            return str(value)
+        return '[]'
+
+    @classmethod
+    def grplist_to_lst(cls, value):
+        """
+        Convert libcomps GrpList to list.
+
+        Args:
+            value: a libcomps GrpList
+
+        Returns:
+            A list
+
+        """
+        grp_list = []
+        if len(value):
+            for i in value:
+                grp_list.append({'name': i.name,
+                                 'default': i.default})
+        return grp_list
+
+    @classmethod
+    def strdict_to_dict(cls, value):
+        """
+        Convert libcomps StrDict type object to standard dict.
+
+        Args:
+            value: a libcomps StrDict
+
+        Returns:
+            lang_dict: a dict
+
+        """
+        lang_dict = {}
+        if len(value):
+            for i, j in value.items():
+                lang_dict[i] = j
+        return lang_dict
+
+    @classmethod
+    def strdict_to_str(cls, value):
```

Comment:
Can you elaborate why it is needed? the conversion to a string version of a dict?
It doesn't seem to be used anywhere.